solutions/USACO jan17-hoof-paper-scissors-gold.mdx

```diff
@@ -0,0 +1,88 @@
+---
+id: usaco-694
+source: USACO Gold 2017 January
+title: Hoof, Paper, Scissors
+author: Neo Wang (C++)
+---
+
+[Official Analysis](http://www.usaco.org/current/data/sol_hps_gold_jan17.html)
+
+## Solution
+
+The main observation for this problem is to realize that we only need to keep track of the number of games played $i$, the number of times switched so far $j$, and the current gesture $k$ in order to determine the most amount of previous games won for any game $i$.
+
+For every move, either Bessie can switch to either `H, P, S` or stay on her current gesture. If she changes her gesture, then the next game $i+1$ will have used $j+1$ gestures, which corresponds to the `dp` state `dp[i+1][j+1][k]`. We can simulate this for all 3 gestures. Then, all we have to do is keep track of whether or not Bessie won at game $i$ with a gesture of $k$. 
+
+Note that you can just compare the current gesture to `H,P,S` because there is always exactly one way to win. 
+
+**Time Complexity:** $\mathcal{O}(NK)$
+
+<LanguageSection>
+
+<CPPSection>
+
+## Implementation
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+using ll = long long;
+using vi = vector<int>;
+using pi = pair<int,int>;
+
+#define sz(x) (int)(x).size()
+#define pb push_back
+#define all(x) begin(x), end(x)
+#define f first
+#define s second
+
+const int MX = 1e5+5;
+
+int dp[MX][25][3]; // dp[i][j][k] is the most amount of games she wins at games i with switches j with current item k
+int A[MX]; // 0 == H 1 == P 2 == S
+
+int main() { 
+    cin.tie(0)->sync_with_stdio(0); 
+    freopen("hps.in", "r", stdin); 
+    freopen("hps.out", "w", stdout);
+
+	int N, K;
+
+	cin >> N >> K;
+
+	for(int i = 0; i < N; i++) { 
+		char a; cin >> a;
+		if(a == 'H') A[i] = 0;
+		if(a == 'P') A[i] = 1;
+		if(a == 'S') A[i] = 2;
+	}
+
+	// either she switches to h or p or s or stays
+
+	for(int i = 0; i <= N; i++) {
+		for(int j = 0; j <= K; j++) {
+			for(int k = 0; k < 3; k++) {
+				if(k == A[i]) dp[i][j][k]++;
+				dp[i+1][j+1][0] = max(dp[i+1][j+1][0], dp[i][j][k]); // switch to not item
+				dp[i+1][j+1][1] = max(dp[i+1][j+1][1], dp[i][j][k]); // switch to not item
+				dp[i+1][j+1][2] = max(dp[i+1][j+1][2], dp[i][j][k]); // switch to not item
```

Comment:
technically, one of these is a "stay"

Reply:
well ig it is a "switch" if you consider that j is being incremented although that's just a simplicity thing since it'll get overwritten nevertheless